source/common/filter/config_discovery_impl.h

```diff
@@ -176,7 +190,61 @@ class HttpDynamicFilterConfigProviderImpl
   Http::FilterFactoryCb instantiateFilterFactory(const Protobuf::Message& message) const override {
     auto* factory = Registry::FactoryRegistry<Server::Configuration::NamedHttpFilterConfigFactory>::
         getFactoryByType(message.GetTypeName());
-    return factory->createFilterFactoryFromProto(message, getStatPrefix(), getFactoryContext());
+    return factory->createFilterFactoryFromProto(message, getStatPrefix(), factory_context_);
+  }
+
+  Server::Configuration::FactoryContext& factory_context_;
+};
+
+// Implementation of a listener dynamic filter config provider.
+template <class FactoryCb>
+class ListenerDynamicFilterConfigProviderImpl : public DynamicFilterConfigProviderImpl<FactoryCb> {
+public:
+  ListenerDynamicFilterConfigProviderImpl(
+      FilterConfigSubscriptionSharedPtr& subscription,
+      const absl::flat_hash_set<std::string>& require_type_urls,
+      Server::Configuration::ListenerFactoryContext& factory_context,
+      ProtobufTypes::MessagePtr&& default_config, bool last_filter_in_filter_chain,
+      const std::string& filter_chain_type, absl::string_view stat_prefix)
+      : DynamicFilterConfigProviderImpl<FactoryCb>(
+            subscription, require_type_urls, factory_context, std::move(default_config),
+            last_filter_in_filter_chain, filter_chain_type, stat_prefix),
+        factory_context_(factory_context) {}
+
+  void validateMessage(const std::string&, const Protobuf::Message&,
+                       const std::string&) const override {}
+
+protected:
+  Server::Configuration::ListenerFactoryContext& factory_context_;
+};
+
+class TcpListenerDynamicFilterConfigProviderImpl
+    : public ListenerDynamicFilterConfigProviderImpl<Network::ListenerFilterFactoryCb> {
+public:
+  using ListenerDynamicFilterConfigProviderImpl::ListenerDynamicFilterConfigProviderImpl;
+
+private:
+  Network::ListenerFilterFactoryCb
+  instantiateFilterFactory(const Protobuf::Message& message) const override {
+    auto* factory =
+        Registry::FactoryRegistry<Server::Configuration::NamedListenerFilterConfigFactory>::
+            getFactoryByType(message.GetTypeName());
+    return factory->createListenerFilterFactoryFromProto(message, nullptr, factory_context_);
```

Comment:
not sure if this has been discussed or not. But indeed I didn't aware from the previous review.

The second parameter of `createListenerFilterFactoryFromProto` is for listener filter matcher
https://github.com/envoyproxy/envoy/blob/54883cf195073d0ba95c992a1196b263c7c748de/source/server/listener_manager_impl.cc#L130-L131

If we just pass the `nullptr` here, then we will lose that feature. https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto#config-listener-v3-listenerfilterchainmatchpredicate

Reply:
Good point! The rest of the occurrence passing nullptr here is because that listener filter is automatically generated, such as adding tls_inspector in some conditions. These code will be killed soon.

In this listener filter using ECDS case, the filter is well-defined and that should work with ECDS. https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener_components.proto#L361-L367

Reply:
Thanks for the comments.

Yes, this is planned in the follow up PR when I working on the item 5, 6, 7 of [issue](https://github.com/envoyproxy/envoy/issues/20049). In this PR,  we  will start to hook the newly added TcpListenerFilterConfigProviderManagerImpl class (in config_discovery_impl.h) into the listener filter factory creation code: https://github.com/envoyproxy/envoy/blob/54883cf195073d0ba95c992a1196b263c7c748de/source/server/listener_manager_impl.cc#L112.

In this function call path:
ProdListenerComponentFactory::createListenerFilterFactoryList_()
  -> FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider()
      -> FilterConfigProviderManagerImpl::createFilterConfigProviderImpl()
         -> ListenerDynamicFilterConfigProviderImpl::ListenerDynamicFilterConfigProviderImpl()

There will be a new parameter: Network::ListenerFilterMatcherSharedPtr (createListenerFilterMatcher(proto_config)) will be propagated down from createListenerFilterFactoryList_() all the way to ListenerDynamicFilterConfigProviderImpl and stored there. And will be used by TcpListenerDynamicFilterConfigProviderImpl::instantiateFilterFactory() during config update.

For UDP listener filter or HTTP filter, a nullptr will be propagated down.

Reply:
Thanks for the comments.

Yes, this is planned in the follow up PR when I working on the item 5, 6, 7 of issue #20049 https://github.com/envoyproxy/envoy/issues/20049 . In this planned PR, we will start to hook the newly added TcpListenerFilterConfigProviderManagerImpl class  into the listener filter factory creation code: https://github.com/envoyproxy/envoy/blob/54883cf195073d0ba95c992a1196b263c7c748de/source/server/listener_manager_impl.cc#L112

In this function call path:

ProdListenerComponentFactory::createListenerFilterFactoryList_()
     ----> FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider() 
        -----> FilterConfigProviderManagerImpl::createFilterConfigProviderImpl()
            -----> ListenerDynamicFilterConfigProviderImpl::ListenerDynamicFilterConfigProviderImpl()

There will be a new parameter: Network::ListenerFilterMatcherSharedPtr (createListenerFilterMatcher(proto_config)), be propagated down from createListenerFilterFactoryList_() all the way to ListenerDynamicFilterConfigProviderImpl and stored there. And will be used by TcpListenerDynamicFilterConfigProviderImpl::instantiateFilterFactory() during config update/removal.

For UDP listener filter or HTTP filter, a nullptr will be propagated down.
Let me add a TODO comment here.